### ISSUE TITLE
Running with coverage is eating variables...

### ISSUE THREAD

**Issue by remy** [tapjs/tapjs#267]
Firstly, yeah, super weird issue subject, not really sure how to retitle it. 

Here's the scenario, when I run `tap index.js` my code passes, when add coverage (any coverage related argument), like `tap index.js --cov` my code fails. I'm using tap@5.7.2 in node@6.0.0.

The `index.js` simply contains this:

``` js
const test = require('tap').test;

test('missing arg', t => {
  var foo = function ({
    user = 'foo',
    ref = 'my-ref',
  } = {}) {
    t.equal(ref, 'my-ref');
    t.end();
  };

  foo({
    user: 'remy',
  });
});
```

When I run with coverage, the value that isn't passed, and should take a default parameter value becomes `undefined`, but **without** the value is correct.

It took me quite a while to par this down, but I've honestly no idea what could be causing it. Any ideas?!!?

**Comment by isaacs** [comment]
My guess would be that Istanbul isn't transpilating this fancy newfangled ES2099 code properly, giving it a bad wrap.

Maybe @gotwarlost or @bcoe might have ideas?

**Comment by bcoe** [comment]
see: https://github.com/bcoe/nyc/issues/265, https://github.com/gotwarlost/istanbul/issues/626, https://github.com/bcoe/nyc/issues/264

tldr; while `istanbul` seems to _mostly_ work for ES2036 there are some issues cropping up ... it feels like someone needs to dive in to `esprima`/`istanbul` and work out some issues.

**Comment by remy** [comment]
This occurred to me too, as I couldn't sleep last night. Sod. I guess I'll have to remove my sparkles in favour of actually coverage reports.

Worth closing the issue or leaving open for others to find until it's fixed deeper down?

**Comment by gotwarlost** [comment]
Sorry guys, I haven't had the time to work on istanbul at all. I'll be in this state for 10 days more and then maybe I'll have some time to look into all the things that I've been ignoring :(

**Comment by bcoe** [comment]
@gotwarlost would love to help in any way I can, including directing folks from `nyc` to the appropriate bits of the Istanbul codebase. To start building out better `ES6` support, any direction as to where to start?

**Comment by bcoe** [comment]
@remy give this a spin:

https://github.com/bcoe/nyc#use-babel-plugin__coverage__-for-better-es6es7-support

we've released better support for ES6/ES7 coverage using the **coverage** instrumentation library.

**Comment by remy** [comment]
Cheers, I'll give it a go tomorrow (enjoying last day of holidays/laptop free!) 😃

**Comment by rmarscher** [comment]
FWIW, I encountered the same issue with default values for destructured arguments not being assigned. It's not in too many places in my code, so I worked around it by checking for `undefined` in the function body. My project is written for node 6.2.x which isn't using babel so it seemed to make sense to change my syntax to what istanbul can handle rather than introduce babel/`__coverage__`. Thanks for all of the work across projects on these ES2015+ issues. 😄

**Comment by isaacs** [comment]
Since this is fundamentally an istanbul/nyc issue, I'm gonna close the issue here.

Please lmk if anything comes up that requires a change in this repo :)

**Comment by remy** [comment]
Finally got around to testing this - just wanted to say all working great, cheers all!

**Comment by remy** [comment]
Urm...I've just been trying this out against my _real_ code and I've found something pretty weird…

I'm not sure whether to comment here, open a new issue…I'm not entirely sure _where_ the problem is too.

Basically, if the required module with the default params is in a outer folder, it doesn't work. 

Yeah, that sounds pretty flipping obscure. Let me know what you'd prefer, I've got a "working" example.

**Comment by remy** [comment]
Following on, without `--cov` it all works okay, so I guess this new thing is against nyc? But…I'm not sure if this tap's usage of nyc. (sorry, rambling, this is a weird one!)

**Comment by isaacs** [comment]
@remy Yeah, you're running into an interesting facade-misdirection issue.  Tap just calls out to nyc, which is a child-process-aware implementation of istanbul's code-wrapping logic.

So I'm almost certain that the issue isn't even nyc, it's istanbul, but it's completely reasonable to assume that tap's to blame, since, ya know, that's _the thing you're using_, of course.

I'm sure that @bcoe and @gotwarlost would <3 a working example posted over at https://github.com/istanbuljs/istanbul-lib-instrument or https://github.com/istanbuljs/nyc
